File: problems/problems_853/solution.py
import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.carFleet(*test_input)

    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        st = []
        for p, s in sorted(zip(position, speed), key=lambda x: x[0]):
            # 避免除法精度运算
            # 左边更快到达的, 一定会和右边更慢到达的形成一个元素, 即单调递增栈
            while st and st[-1][0] * s <= (target - p) * st[-1][1]:
                st.pop()
            st.append(((target - p), s))
        return len(st)

File: problems/problems_853/Solution.java
package problems.problems_853;

import com.alibaba.fastjson.JSON;
import java.util.*;
import qubhjava.BaseSolution;


public class Solution extends BaseSolution {
    public int carFleet(int target, int[] position, int[] speed) {
        
    }

    @Override
    public Object solve(String[] inputJsonValues) {
        int target = Integer.parseInt(inputJsonValues[0]);
		int[] position = jsonArrayToIntArray(inputJsonValues[1]);
		int[] speed = jsonArrayToIntArray(inputJsonValues[2]);
        return JSON.toJSON(carFleet(target, position, speed));
    }
}
